terminology/inmemory_terminology.bal

```diff
@@ -0,0 +1,439 @@
+    }
+
+    // Define the populateCodeSystemMap function
+    isolated function populateCodeSystemMap(json internalFhirCodeSystems, json externalFhirCodeSystems, map<r4:CodeSystem> codeSystemMap) returns map<r4:CodeSystem> {
+        // Implementation logic goes here
+        map<r4:CodeSystem> resultedCodeSystemMap = codeSystemMap;
+        // Populate the resultedCodeSystemMap with the code systems from internalFhirCodeSystems and externalFhirCodeSystems
+        foreach json codeSystem in [internalFhirCodeSystems, externalFhirCodeSystems] {
+            foreach json jCodeSystem in <json[]>codeSystem {
+                r4:CodeSystem|error c = jCodeSystem.cloneWithType();
+                if c is error {
+                    r4:FHIRError fHIRError = r4:createFHIRError(
+                                            "Error occurred while type casting json code system to CodeSystem type", r4:ERROR,
+                                            r4:PROCESSING,
+                                            diagnostic = c.message(),
+                                            cause = c
+                                        );
+                    log:printError(fHIRError.toBalString());
+                } else {
+                    string key = getKey(<string>c.url, <string>c.version);
+                    resultedCodeSystemMap[key] = c;
+                }
+            }
+        }
+
+        // Return the updated codeSystemMap
+        return resultedCodeSystemMap;
+    }
+
+    isolated function populateValueSetMap(json internalFhirValueSets, json externalFhirValueSets, map<r4:ValueSet> valueSetMap) returns map<r4:ValueSet> {
+        // Implementation logic goes here
+        map<r4:ValueSet> resultedValueSetMap = valueSetMap;
+        // Populate the codeSystemMap with the code systems from internalFhirCodeSystems and externalFhirCodeSystems
+        foreach json valueSet in [internalFhirValueSets, externalFhirValueSets] {
+            foreach json jValueSet in <json[]>valueSet {
+                r4:ValueSet|error v = jValueSet.cloneWithType();
+                if v is error {
+                    r4:FHIRError fHIRError = r4:createFHIRError(
+                                            "Error occurred while type casting json value set to ValueSet type", r4:ERROR,
+                                            r4:PROCESSING,
+                                            diagnostic = v.message(),
+                                            cause = v
+                                        );
+                    log:printError(fHIRError.toBalString());
+                } else {
+                    string key = getKey(<string>v.url, <string>v.version);
+                    resultedValueSetMap[key] = v;
+                }
+            }
+        }
+
+        // Return the updated codeSystemMap
+        return resultedValueSetMap;
+    }
+
+    public isolated function addCodeSystem(r4:CodeSystem codeSystem) returns r4:FHIRError? {
```

Comment:
Let's add bal docs for public functions

Reply:
This is inherited from the Terminology interface.